### PR TITLE
Support for adding patterns and receiving knock data from the Android app

### PR DESCRIPTION
**Pull request by JakeDarling**
POST /patterns/ to add new AccessPattern. It gets a field called 'pending' set to True

POST /knock/ from Android. If an AccessPattern is pending, it gets its PatternPieces set

### CODE REVIEW

File: knockserver/views.py [hunk 1]
```diff
@@ -6,8 +6,83 @@
...
 from knockserver.database import db_session
-from knockserver.models import ExampleObject
+from knockserver.models import AccessPattern, PatternPiece
 
 @app.route('/')
 def index():
     return render_template('index.html')
+
+@app.route('/patterns/', methods=['GET', 'POST'])
+def patterns():
```
Comment (corbanmailloux) [inline knockserver/views.py]: To increase readability, I'd suggest breaking this into two functions. One for the `GET` (`def patterns_get()`) and one for the `POST` (`def patterns_post()`). Then it's one less `if` we need, and it moves that logic to the routing controller.

File: knockserver/views.py [hunk 2]
```diff
@@ -6,8 +6,83 @@
...
+    # Check for pending knock
+    pending_pattern = AccessPattern.query.filter(AccessPattern.pending == True).first()
+
+    if pending_pattern is not None:
+        i = 0
+        for json_val in data['pattern']:
+            piece = PatternPiece()
+            piece.length = data['pattern'][i]
+            piece.order = i
+            #TODO Don't remember if first is pressed
```
Comment (corbanmailloux) [inline knockserver/views.py]: What does this `TODO` mean?

**Review by TempusShift — APPROVED** [review]
Looks good!

**Review by kmb4577 — APPROVED** [review]

**Review by ip2176 — APPROVED** [review]

File: knockserver/views.py [hunk 3]
```diff
@@ -6,8 +6,90 @@
...
+    return '{"success": true}' #TODO decide response
+
+
+@app.route('/patterns/', methods = ['GET'])
+def patterns_get():
+    """
+    Return all patterns for debugging
+    Patterns ordered newest first
+    """
+    return str(AccessPattern.query.order_by(AccessPattern.id.desc().all()))
```
Comment (corbanmailloux) [inline knockserver/views.py]: I haven't run this line to test it, but does this return JSON? I think we should generally try to JSON-ify everything.

**Review by corbanmailloux — COMMENTED** [review]
(This is not a blocker for this PR, but I think we should talk about it soon.)
How do people feel about enforcing a style guide for this project? Because it's Python, I'd suggest something like Flake8.

File: knockserver/views.py (hunk 3 above)
Comment (corbanmailloux) [reply knockserver/views.py]: Also, the parentheses are incorrect on this. `.all()` should be called on `query`.
I'm about to push a fix for that.

**Pull request merged**